### ISSUE TITLE
How to extract trained binarized weights with DoReFa?

### ISSUE THREAD

**Issue by zhouchch3** [tensorpack/tensorpack#1199]
Why the trained models (eg. 1, 2,6) contain only the float32 weights in each convolution layer instead of quantized 1-bit-number ones?

**Comment by ppwwyyxx** [comment]
Duplicate of https://github.com/tensorpack/tensorpack/issues/573

**Comment by zhouchch3** [comment]
Thank you very much!
But I get another problem. I run fw to get the actual weights ( bitW = 1 )
I get -0.2501563 and 0.2501563 as follow:
![image](https://user-images.githubusercontent.com/33385095/58007556-1e0dce00-7b1d-11e9-8532-ece3c3c36fba.png)
I want to know whether 0.2501563  means 1 and -0.2501563  means -1?? @ppwwyyxx

**Comment by ppwwyyxx** [comment]
No. It means 0.2501563 and -0.2501563. Although in a binarized implementation you'll want to use 1 and -1 to represent them.

**Comment by zhouchch3** [comment]
> No. It means 0.2501563 and -0.2501563. Although in a binarized implementation you'll want to use 1 and -1 to represent them.

So how can I get binarized weights like 1 and -1?  Or can I directly use 1 and -1 to represent 0.2501563 and -0.2501563?

**Comment by ppwwyyxx** [comment]
0.2501563 and -0.2501563 are the binarized weights and this is what "binarized" means in all the binarized networks. You can use 1 and -1 as long as you compensate for the scale at a different place.

**Comment by zhouchch3** [comment]
@ppwwyyxx 
**When I try to run AlenNet pretrained model like this:**

![image](https://user-images.githubusercontent.com/33385095/58017839-cc247280-7b33-11e9-813d-ef04ffab3d18.png)

**I get this problem as follows:**
![image](https://user-images.githubusercontent.com/33385095/58017890-f0804f00-7b33-11e9-8408-6ad32c38dfee.png)

**What is wrong? Thank you**

**Comment by ppwwyyxx** [comment]
The code will download some label files to `tensorpack_data` (full path shown in your picture) and it seems the download is broken. Please remove any files in the above directory, check your internet connection and try again.

**Comment by zhouchch3** [comment]
@ppwwyyxx  Thank you!
What do I have to modify to train AlexNet on **Tiny-ImageNet** instead of ImageNet?
`Tiny Imagenet has only 200 classes. Each class has 500 training images, 50 validation images, and 50 test images.`
Another question is whether I can use DoreFa to train AlexNet with  (W, A) = ( 1, 8 )?

**Comment by ppwwyyxx** [comment]
> What do I have to modify

You have to implement the data loader for a different dataset, and change the number of classes in the last FC layer.

> Another question is whether I can use DoreFa to train AlexNet with (W, A) = ( 1, 8 )?

You can.

**Comment by zhouchch3** [comment]
@ppwwyyxx 

> You have to implement the data loader for a different dataset, and change the number of classes in the last FC layer.

I will try it. Thank you very much

**Comment by zhouchch3** [comment]
@ppwwyyxx 
I try to train use `./alexnet-dorefa.py --dorefa 1,2,6 --data PATH --gpu 0,1
`
I get this error:
![image](https://user-images.githubusercontent.com/33385095/58297905-1778a280-7e0c-11e9-8d85-56cf28665a84.png)

I have installed tensorflow-gpu and Cuda-10.0 but **not** installed CuDNN. If I install CuDNN, this problem will be solved?

**Comment by ppwwyyxx** [comment]
For anyone to better understand and diagnose your issue, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))
Include full logs and environment info.
